packets/packets.go

```diff
@@ -484,84 +554,54 @@ type headPayloadShape struct {
 	Sizes []int16
 }
 
-// readTLV reads data for size bytes, generating a list of all TLV objects
-func readTLV(data io.Reader, size uint8) (result []interface{}, err error) {
-	var t uint8
-	var tlvsize uint8
-	for size > 0 {
-		if size < 8 {
-			return result, fmt.Errorf("readTLV needs to read multiples of 8 bytes")
-		}
-		if err = binary.Read(data, binary.BigEndian, &t); err != nil {
-			return result, err
+// parseTLV parses data, generating a list of all TLV objects
+func parseTLV(data []byte) (result []interface{}, err error) {
+	bytesRemaining := len(data)
+	for bytesRemaining > 0 {
+		if bytesRemaining < 8 {
+			return result, fmt.Errorf("parseTLV needs to read multiples of 8 bytes")
 		}
-		if err = binary.Read(data, binary.BigEndian, &tlvsize); err != nil {
-			return result, err
+		t := data[0]
+		tlvsize := 8*int(data[1])
+		if tlvsize > bytesRemaining {
+			return result, fmt.Errorf("TLV type 0x%x has len %d, but remaining hdr size is %d",
+				t, tlvsize, bytesRemaining)
 		}
-		if 8*tlvsize > size {
-			return result, fmt.Errorf("TLV type 0x%x has len 8*%d, but remaining hdr size is %d",
-				t, tlvsize, size)
+		if tlvsize <= 0 {
+			return result, fmt.Errorf("TLV reports negative size %d", tlvsize)
 		}
 		switch t {
 		case tlvTAG:
-			var x uint16
-			var tag PacketTag
-			if err = binary.Read(data, binary.BigEndian, &x); err != nil {
-				return result, err
-			}
+			x := binary.BigEndian.Uint16(data[2:])
+			tag := PacketTag(binary.BigEndian.Uint32(data[4:]))
 			if x != 0 {
 				return result, fmt.Errorf("TAG TLV has value 0x%x, expect 0", x)
 			}
-			if err = binary.Read(data, binary.BigEndian, &tag); err != nil {
-				return result, err
-			}
 			result = append(result, tag)
 
 		case tlvTIMESTAMP: // timestamps without units
-			var x uint16
-			var y uint32
-			if err = binary.Read(data, binary.BigEndian, &x); err != nil {
-				return result, err
-			}
-			if err = binary.Read(data, binary.BigEndian, &y); err != nil {
-				return result, err
-			}
+			x := binary.BigEndian.Uint16(data[2:])
```

Comment:
Why are we using binary here? is it because the `getbytes` style approach only works on the native endianness, which is little? If so I think its fine. If not, I'd be in favor of switching to a getbytes approach.

Reply:
It's what you said: that we need to handle big endian, which `binary` just does. That package is slow for slices, but this is actually grabbing a single `uint16` value. Honestly, I wish we weren't programming the packets to use big endian numbers in the packet headers, but we are. The firmware team basically called that a "micro optimization" and blew me off. Seems to me it would have considerably optimized MY time to use little endian, but possibly the reverse is true on the firmware end???

Sure, I could add byte swapping to `getbytes`, but that seems unlikely to be faster, and it duplicates the widely-used, correct (if slow) code that already appears in the standard `binary` package.